File: 미로 경로 출력하기/solution_ac.cpp
#include <iostream>

int N, M;
char map[2020][2020];
bool visited[2020][2020];

int dx[4] = { 0, 0, 1, -1 };
int dy[4] = { 1, -1, 0, 0 };

void init();
void solve();
void print();

int main(void)
{
    std::cin.tie(0);
    std::ios_base::sync_with_stdio(false);

    init();
    solve();
    print();

    return 0;
}

void init() {
    std::cin >> N >> M;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            std::cin >> map[i][j];
            if(map[i][j] == '.') map[i][j] = '@';
        }
    }
}

void find_start_end_points(int& start_x, int& start_y, int& end_x, int& end_y) {
    for (int i = 0; i < M; i++) {
        if (map[0][i] == '@') {
            if (start_x == -1) {
                start_x = 0;
                start_y = i;
            } else {
                end_x = 0;
                end_y = i;
            }
        }
        if (map[N - 1][i] == '@') {
            if (start_x == -1) {
                start_x = N - 1;
                start_y = i;
            } else {
                end_x = N - 1;
                end_y = i;
            }
        }
    }
    for (int i = 0; i < N; i++) {
        if (map[i][0] == '@') {
            if (start_x == -1) {
                start_x = i;
                start_y = 0;
            } else {
                end_x = i;
                end_y = 0;
            }
        }
        if (map[i][M - 1] == '@') {
            if (start_x == -1) {
                start_x = i;
                start_y = M - 1;
            } else {
                end_x = i;
                end_y = M - 1;
            }
        }
    }
}

bool dfs(int si, int sj, int ei, int ej) {
    if (si == ei && sj == ej) {
        map[si][sj] = '.';
        return true;
    }

    bool res = false;

    for (int dir = 0; dir < 4; dir++) {
        int nx = si + dx[dir];
        int ny = sj + dy[dir];

        if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
        if (map[nx][ny] == '+') continue;
        if (visited[nx][ny]) continue;

        visited[nx][ny] = true;
        res = dfs(nx, ny, ei, ej);
        if (res) {
            map[si][sj] = '.';
            return true;
        }
    }

    return false;
}

void solve() {
    int si = -1, sj = -1, ei = -1, ej = -1;
    find_start_end_points(si, sj, ei, ej);

    dfs(si, sj, ei, ej);
}

void print() {
    for (int i = 0; i < N;i++) {
        for (int j = 0; j < M; j++) {
            std::cout << map[i][j];
        }
        std::cout << "\n";
    }
}

File: 미로 경로 출력하기/solution_ac.py
import sys
sys.setrecursionlimit(4040404)

def init():
    global N, M, maze, visited, dx, dy
    N, M = map(int, input().split())
    maze = [list(input().replace('.','@')) for _ in range(N)]
    visited = [[False] * M for _ in range(N)]
    dx = [0, 1, 0, -1]
    dy = [1, 0, -1, 0]


def dfs(si, sj):
    if si == ei and sj == ej:
        maze[si][sj] = '.'
        return True

    res = False

    for i in range(4):
        nx = si + dx[i]
        ny = sj + dy[i]

        if nx < 0 or nx >= N or ny < 0 or ny >= M:
            continue
        if maze[nx][ny] == '+':
            continue
        if visited[nx][ny]:
            continue

        visited[nx][ny] = True
        res = dfs(nx, ny)
        if res:
            maze[si][sj] = '.'
            return True

    return False


def solve():
    global ei,ej
    si, sj, ei, ej = -1, -1, -1, -1
    for i in range(M):
        if maze[0][i] == '@':
            if si == -1:
                si, sj = 0, i
            else:
                ei, ej = 0, i
        if maze[N - 1][i] == '@':
            if si == -1:
                si, sj = N - 1, i
            else:
                ei, ej = N - 1, i
    for i in range(N):
        if maze[i][0] == '@':
            if si == -1:
                si, sj = i, 0
            else:
                ei, ej = i, 0
        if maze[i][M - 1] == '@':
            if si == -1:
                si, sj = i, M - 1
            else:
                ei, ej = i, M - 1
    
    dfs(si, sj)

def _print():
    t = []
    for i in range(N):
        t.append(''.join(maze[i]))
    print('\n'.join(t))

if __name__ == '__main__':
    init()
    solve()
    _print()
